File: workers/realtime-camera-worker/src/ppe_registry.py
# MIA — KKD Ekipman Kaydı (worker tarafı) — js/ppe-registry.js ile SENKRON tutun.
# DÜRÜSTLÜK: yalnız modelin (rf-27) gerçekten içerdiği sınıflar 'supported/experimental'.
# requires_training ekipmanlar worker'da ASLA olay üretmez (model sınıfı yok).

REGISTRY = {
    "helmet": {
        "status": "supported", "violation_class": "NO-Hardhat", "ok_class": "Hardhat",
        "violation_type": "no_helmet", "default_risk": "high",
    },
    "safety_vest": {
        "status": "supported", "violation_class": "NO-Safety Vest", "ok_class": "Safety Vest",
        "violation_type": "no_vest", "default_risk": "high",
    },
    "mask": {
        "status": "experimental", "violation_class": "NO-Mask", "ok_class": "Mask",
        "violation_type": "no_mask", "default_risk": "medium",
    },
    # requires_training — model sınıfı YOK, olay üretilemez:
    "gloves": {"status": "requires_training", "violation_class": None, "ok_class": None,
               "violation_type": "no_gloves", "default_risk": "medium"},
    "safety_glasses": {"status": "requires_training", "violation_class": None, "ok_class": None,
                       "violation_type": "no_glasses", "default_risk": "medium"},
    "safety_harness": {"status": "requires_training", "violation_class": None, "ok_class": None,
                       "violation_type": "no_harness", "default_risk": "critical"},
    "safety_boots": {"status": "requires_training", "violation_class": None, "ok_class": None,
                     "violation_type": "no_boots", "default_risk": "low"},
    "ear_protection": {"status": "requires_training", "violation_class": None, "ok_class": None,
                       "violation_type": "no_ear_protection", "default_risk": "low"},
}

VALID_RISKS = ("low", "medium", "high", "critical")
DEFAULT_REQUIRED = {"helmet": True, "safety_vest": True, "mask": False}


def build_violation_map(required_equipment, risk_rules=None):
    """Profilin etkin ekipmanından model-sınıfı → (violation_type, risk) haritası üretir.

    required_equipment: {"helmet": true, "safety_vest": true, ...} (profilden)
    risk_rules:         {"helmet": "high", ...} (opsiyonel geçersiz kılma)
    requires_training ekipmanlar SESSİZCE atlanır (tespit edilemez — dürüstlük).
    """
    req = required_equipment if isinstance(required_equipment, dict) else DEFAULT_REQUIRED
    rules = risk_rules if isinstance(risk_rules, dict) else {}
    vmap = {}
    for key, enabled in req.items():
        if not enabled:
            continue
        item = REGISTRY.get(key)
        if not item or not item["violation_class"]:
            continue  # desteklenmeyen ekipman: olay üretme
        risk = rules.get(key) or item["default_risk"]
        if risk not in VALID_RISKS:
            risk = item["default_risk"]
        vmap[item["violation_class"]] = (item["violation_type"], risk, key)
    return vmap


def equipment_summary(required_equipment, predictions):
    """Olay kaydı için detected/missing/required ekipman özetleri (sınıf adlarından)."""
    req = required_equipment if isinstance(required_equipment, dict) else DEFAULT_REQUIRED
    # Eski biçim ("class") ve Faz 15 normalize biçim ("raw_class_name") desteklenir
    classes = {p.get("class") or p.get("raw_class_name") for p in (predictions or [])}
    detected, missing = {}, {}
    for key, enabled in req.items():
        item = REGISTRY.get(key)
        if not enabled or not item or not item["violation_class"]:
            continue
        if item["ok_class"] in classes:
            detected[key] = True
        if item["violation_class"] in classes:
            missing[key] = True
    return {"required": {k: v for k, v in req.items() if v}, "detected": detected, "missing": missing}

File: js/ppe-registry.js
// === MIA — Türkiye İnşaat KKD Ekipman Kaydı (Faz 13) ===
// TEK GERÇEK KAYNAK: hangi ekipman taranabilir, hangi model sınıfıyla, hangi durumda.
// Worker tarafı eşdeğeri: workers/realtime-camera-worker/src/ppe_registry.py (senkron tutun).
//
// DÜRÜSTLÜK KURALI: status alanı modelin (construction-site-safety/27 = rf-27)
// GERÇEK yeteneğini yansıtır. rf-27 sınıfları: Hardhat, NO-Hardhat, Safety Vest,
// NO-Safety Vest, Mask, NO-Mask, Person, Safety Cone, machinery, vehicle.
// Desteklenmeyen ekipman "requires_training" işaretlenir ve ETKİNLEŞTİRİLEMEZ —
// sahte tespit vaadi yoktur. Model eğitilince burada (ve .py'de) terfi ettirilir.
//
//   supported         → şimdi tespit edilir (modelde NO-* ihlal sınıfı var)
//   experimental      → sınıf modelde var ama saha doğrulaması yetersiz
//   requires_training → modelde sınıf YOK; eğitim verisi gerekir (UI'da kilitli)
(function () {
    var REGISTRY = [
        { key: "helmet",         label_tr: "Baret",            label_en: "Hard hat",
          status: "supported",   violation_class: "NO-Hardhat",     ok_class: "Hardhat",
          violation_type: "no_helmet", default_risk: "high",
          note_tr: "rf-27 ile tespit edilir; saha doğrulaması pilotta ölçülür." },
        { key: "safety_vest",    label_tr: "Reflektörlü Yelek", label_en: "Safety vest",
          status: "supported",   violation_class: "NO-Safety Vest", ok_class: "Safety Vest",
          violation_type: "no_vest", default_risk: "high",
          note_tr: "rf-27 ile tespit edilir; saha doğrulaması pilotta ölçülür." },
        { key: "mask",           label_tr: "Maske / Solunum Koruma", label_en: "Mask / respirator",
          status: "experimental", violation_class: "NO-Mask",       ok_class: "Mask",
          violation_type: "no_mask", default_risk: "medium",
          note_tr: "Modelde sınıf var; inşaat sahasında yanlış pozitif oranı henüz ölçülmedi." },
        { key: "gloves",         label_tr: "Eldiven",          label_en: "Gloves",
          status: "requires_training", violation_class: null, ok_class: null,
          violation_type: "no_gloves", default_risk: "medium",
          note_tr: "rf-27 bu sınıfı içermiyor — model eğitimi gerekir (yol haritası)." },
        { key: "safety_glasses", label_tr: "Koruyucu Gözlük",  label_en: "Safety glasses",
          status: "requires_training", violation_class: null, ok_class: null,
          violation_type: "no_glasses", default_risk: "medium",
          note_tr: "rf-27 bu sınıfı içermiyor — model eğitimi gerekir." },
        { key: "safety_harness", label_tr: "Emniyet Kemeri",   label_en: "Safety harness",
          status: "requires_training", violation_class: null, ok_class: null,
          violation_type: "no_harness", default_risk: "critical",
          note_tr: "Yüksekte çalışma için kritik — özel eğitim verisi gerekir." },
        { key: "safety_boots",   label_tr: "İş Ayakkabısı",    label_en: "Safety boots",
          status: "requires_training", violation_class: null, ok_class: null,
          violation_type: "no_boots", default_risk: "low",
          note_tr: "rf-27 bu sınıfı içermiyor — model eğitimi gerekir." },
        { key: "ear_protection", label_tr: "Kulak Koruyucu",   label_en: "Ear protection",
          status: "requires_training", violation_class: null, ok_class: null,
          violation_type: "no_ear_protection", default_risk: "low",
          note_tr: "rf-27 bu sınıfı içermiyor — model eğitimi gerekir." }
    ];
    // Bağlam sınıfları (ekipman değil; tespit/ileri analitik için):
    var CONTEXT_CLASSES = ["Person", "Safety Cone", "machinery", "vehicle"];

    window.MIAPpeRegistry = {
        all: function () { return REGISTRY.slice(); },
        get: function (key) {
            for (var i = 0; i < REGISTRY.length; i++) if (REGISTRY[i].key === key) return REGISTRY[i];
            return null;
        },
        selectable: function () {  // UI'da etkinleştirilebilir olanlar
            return REGISTRY.filter(function (r) { return r.status !== "requires_training"; });
        },
        contextClasses: CONTEXT_CLASSES,
        statusLabel: function (s) {
            return s === "supported" ? "Destekleniyor"
                 : s === "experimental" ? "Deneysel"
                 : "Model eğitimi gerekir";
        }
    };
})();
